### DISCUSSION TITLE
Setting PID control rules with pyswmm

### DISCUSSION THREAD

**Discussion by GuillaumeRammal** [pyswmm/pyswmm#380] [Q&A] [answered]
Hello everbody,

Has anyone tried to set PID control rules for an orifice or a mobile weir through pyswmm ? or even edited the PID controls already coded in the SWMM file ?

I am trying to acces the control rules in the swmm file to modify the PId parameters, but I don't know how to do it.

This is part of my code:

_with Simulation('Test_model.inp') as sim:
    link_list = Links(sim)
    Rule1_link = link_list['C1']
    for step in sim:
        if Rule1_link.flow != 1:
            Rule1_link.target_setting = ['PID 0.1 1000.0 0.0']_ --> But here I should enter a float variable not an array

Could anyone help me please?

**Comment by bemcdonnell** [comment, marked as answer]
@GuillaumeRammal,

The target setting only accept a float. It isn’t like a normal control rule in SWMM. If you want to use a PID, you’d need to write your own PID function to return the one control number. Your next best option might be to implement a control curve function.  I’m traveling and the moment - I can provide an example later

**Reply by GuillaumeRammal** [reply]
Thank you so much Bryant !! I'll try the 2 options although the first one of coding the PID function seems a little bit tricky.
If anyone has written such a function to control the PID parameters, I would be so grateful for sharing it with me!!

**Reply by bemcdonnell** [reply]
@GuillaumeRammal just a hint, here is SWMM’s PID 

https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/7e7e6da799e4c8f1fd51499cebe47f82f80c67f6/src/solver/controls.c#L819

**Reply by GuillaumeRammal** [reply]
Thank you again <3 <3.
I will check it  out, but I might still need your help in incorporating this function in the pyswmm tool.
